src/tmt500rs/hid-tmt500rs-usb.c

```diff
@@ -0,0 +1,1260 @@
+  if (ret) {
+    hid_err(t500rs->hdev, "Failed to send Report 0x02: %d\n", ret);
+    return ret;
+  }
+
+  /* Report 0x01 - Main effect upload for periodic (set waveform/type) */
+  memset(buf, 0, 15);
+  buf[0] = 0x01;
+  buf[1] = 0x00;        /* Effect ID 0 */
+  buf[2] = effect_type; /* Waveform type (0x20..0x24) */
+  buf[3] = 0x40;
+  buf[4] = 0xff;
+  buf[5] = 0xff;
+  buf[6] = 0x00;
+  buf[7] = 0xff;
+  buf[8] = 0xff;
+  buf[9] = 0x0e; /* Parameter subtype reference */
+  buf[10] = 0x00;
+  buf[11] = 0x1c; /* Envelope subtype reference */
+  buf[12] = 0x00;
+  buf[13] = 0x00;
+  buf[14] = 0x00;
+  ret = t500rs_send_usb(t500rs, buf, 15);
+  if (ret) {
+    hid_err(t500rs->hdev, "Failed to send Report 0x01 (periodic main): %d\n",
+            ret);
+    return ret;
+  }
+
+  /* Report 0x04 - Periodic parameters */
+  memset(buf, 0, 15);
+  buf[0] = 0x04;
+  buf[1] = 0x0e;
+  buf[2] = 0x00;
+  buf[3] = mag;                  /* Magnitude */
+  buf[4] = 0x00;                 /* Offset */
+  buf[5] = 0x00;                 /* Phase */
+  buf[6] = period & 0xff;        /* Period low byte */
+  buf[7] = (period >> 8) & 0xff; /* Period high byte */
+  ret = t500rs_send_usb(t500rs, buf, 8);
+  if (ret) {
+    hid_err(t500rs->hdev, "Failed to send Report 0x04: %d\n", ret);
+    return ret;
+  }
+
+  /* Report 0x01 - Main effect upload */
+  memset(buf, 0, 15);
+  buf[0] = 0x01;
+  buf[1] = effect->id;
+  buf[2] = effect_type; /* Waveform type */
+  buf[3] = 0x40;
+  buf[4] = 0x17;
+  buf[5] = 0x25;
+  buf[6] = 0x00;
+  buf[7] = 0xff;
+  buf[8] = 0xff;
+  buf[9] = 0x0e;
+  buf[10] = 0x00;
+  buf[11] = 0x1c;
+  buf[12] = 0x00;
+  buf[13] = 0x00;
+  buf[14] = 0x00;
+  ret = t500rs_send_usb(t500rs, buf, 15);
+  if (ret) {
+    hid_err(t500rs->hdev, "Failed to send Report 0x01: %d\n", ret);
+    return ret;
+  }
+
+  T500RS_DBG("%s effect %d uploaded\n", type_name, effect->id);
+  return 0;
+}
+
+/* Upload ramp effect */
+static int t500rs_upload_ramp(struct t500rs_device_entry *t500rs,
+                              const struct tmff2_effect_state *state) {
+  const struct ff_effect *effect = &state->effect;
+  u8 *buf = t500rs->send_buffer; /* Use DMA-safe buffer */
+  int ret;
+  int start_level = effect->u.ramp.start_level;
+  int end_level = effect->u.ramp.end_level;
+  u16 duration_ms = effect->replay.length;
+  u16 start_scaled;
+
+  /* Scale to 0-255 */
+  start_scaled = (abs(start_level) * 0xff) / 32767;
+
+  T500RS_DBG("Upload ramp: id=%d, start=%d, end=%d, duration=%dms\n",
+             effect->id, start_level, end_level, duration_ms);
+
+  /* Report 0x02 - Envelope */
+  memset(buf, 0, 15);
+  buf[0] = 0x02;
+  buf[1] = 0x1c;
+  buf[2] = 0x00;
+  ret = t500rs_send_usb(t500rs, buf, 9);
+  if (ret) {
+    hid_err(t500rs->hdev, "Failed to send Report 0x02: %d\n", ret);
+    return ret;
+  }
+
+  /* Report 0x04 - Ramp parameters */
+  /* NOTE: T500RS doesn't support native ramp - just holds start level */
```

Comment:
What does that mean? What is a "native" ramp? I would expect that the wheel handles the ramp effect, how did you determined that the ramp is not supported?